pkg/tecdsa/signing/protocol_test.go

```diff
@@ -0,0 +1,302 @@
+package signing
+
+import (
+	"fmt"
+	"github.com/keep-network/keep-core/pkg/crypto/ephemeral"
+	"github.com/keep-network/keep-core/pkg/internal/testutils"
+	"github.com/keep-network/keep-core/pkg/protocol/group"
+	"reflect"
+	"testing"
+)
+
+// TODO: This file contains unit tests that stress each protocol phase
+//       separately. We should also develop integration tests checking the
+//       whole signing protocol.
+
+const (
+	groupSize          = 3
+	dishonestThreshold = 0
+	sessionID          = "session-1"
+)
+
+func TestGenerateEphemeralKeyPair(t *testing.T) {
+	members, err := initializeEphemeralKeyPairGeneratingMembersGroup(
+		dishonestThreshold,
+		groupSize,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Generate ephemeral key pairs for each group member.
+	messages := make(map[group.MemberIndex]*ephemeralPublicKeyMessage)
+	for _, member := range members {
+		message, err := member.generateEphemeralKeyPair()
+		if err != nil {
+			t.Fatal(err)
+		}
+		messages[member.id] = message
+	}
+
+	// Assert that each member has a correct state.
+	for _, member := range members {
+		// Assert the right key pairs count is stored in the member's state.
+		expectedKeyPairsCount := groupSize - 1
+		actualKeyPairsCount := len(member.ephemeralKeyPairs)
+		testutils.AssertIntsEqual(
+			t,
+			fmt.Sprintf(
+				"number of stored ephemeral key pairs for member [%v]",
+				member.id,
+			),
+			expectedKeyPairsCount,
+			actualKeyPairsCount,
+		)
+
+		// Assert the member does not hold a key pair with itself.
+		_, ok := member.ephemeralKeyPairs[member.id]
+		if ok {
+			t.Errorf(
+				"[member:%v] found ephemeral key pair generated to self",
+				member.id,
+			)
+		}
+
+		// Assert key pairs are non-nil.
+		for otherMemberID, keyPair := range member.ephemeralKeyPairs {
+			if keyPair.PrivateKey == nil {
+				t.Errorf(
+					"[member:%v] key pair's private key not set for member [%v]",
+					member.id,
+					otherMemberID,
+				)
+			}
+
+			if keyPair.PublicKey == nil {
+				t.Errorf(
+					"[member:%v] key pair's public key not set for member [%v]",
+					member.id,
+					otherMemberID,
+				)
+			}
+		}
+	}
+
+	// Assert that each message is formed correctly.
+	for memberID, message := range messages {
+		// We should always be the sender of our own messages.
+		testutils.AssertIntsEqual(
+			t,
+			"message sender",
+			int(memberID),
+			int(message.senderID),
+		)
+
```

Comment:
Should we assert `groupSize - 1` public keys are generated in each message?

Reply:
Addressed in https://github.com/keep-network/keep-core/pull/3228/commits/c0fca3b97e2908eb5d6956a825aa141836d8951a